arch/arm/include/aarch32/cortex_m/wrap_func.h

```diff
@@ -0,0 +1,131 @@
+ */
+#define WRAP_FUNC_RAW(preface, name, postface, store_lr, load_lr) \
+	do { \
+		__asm(".global "#preface"; .type "#preface", %function"); \
+		__asm(".global "#name"; .type "#name", %function"); \
+		__asm(".global "#postface"; .type "#postface", %function"); \
+		\
+		__asm(store_lr); \
+		__asm("push {r0-r3}"); \
+		__asm("bl "#preface); \
+		__asm("pop {r0-r3}"); \
+		__asm("bl "#name); \
+		__asm("push {r0-r3}"); \
+		__asm("bl "#postface); \
+		__asm("pop {r0-r3}"); \
+		__asm(load_lr); \
+	} while (0)
+
+/**
+ * @brief Macro for "sandwiching" a function call (@p name) in two other calls
+ *
+ * This is for functions which do not pass arguments or return values via the
+ * stack. I.e. the arguments and return values must each fit within 4 words,
+ * after accounting for alignment.
+ *
+ * If nothing is passed on the stack, the stack can be used to store LR without
+ * messing with offsets.
+ *
+ *	int foo(char *arg); // Implemented elsewhere.
+ *	int __attribute__((naked)) foo_wrapped(char *arg)
+ *		{WRAP_FUNC(bar, foo, baz)}
+ *
+ * is equivalent to
+ *
+ *	int foo(char *arg); // Implemented elsewhere.
+ *	int foo_wrapped(char *arg)
+ *	{
+ *		bar();
+ *		int res = foo(arg);
+ *		baz();
+ *		return res;
+ *	}
+ *
+ * @note __attribute__((naked)) is not mandatory, but without it, GCC gives a
+ *       warning for functions with a return value. It also saves a bit of space
+ *       since it removes a little code that is not necessary.
+ *
+ * See @ref WRAP_FUNC_RAW for more information.
+ */
+#define WRAP_FUNC(preface, name, postface) \
+	WRAP_FUNC_RAW(preface, name, postface, "push {lr}", "pop {pc}")
+
+/**
+ * @brief Macro for "sandwiching" a function call (@p name) in two other calls
+ *
+ * This is for functions which pass arguments or return values via the stack.
+ * I.e. the arguments or return values exceed 4 words, after accounting for
+ * alignment.
+ *
+ * This means LR cannot be stored on the stack, so it must be stored in an extra
+ * argument called lr_bak, see example below.
+ *
+ *	int foo(char *arg1, char *arg2, int arg3, uint64 arg4);
+ *	int __attribute__((naked)) foo_wrapped(char *arg1, char *arg2, int arg3,
+ *						uint64 arg4, uint32_t lr_bak)
+ *		{WRAP_FUNC_STACK_ARGS(bar, foo, baz)}
+ *
+ * is equivalent to
+ *
+ *	int foo(char *arg1, char *arg2, int arg3, uint64 arg4);
+ *	int foo_wrapped(char *arg1, char *arg2, int arg3, uint64 arg4,
+ *			uint32_t lr_bak)
+ *	{
+ *		bar();
+ *		int res = foo(arg1, arg2, arg3, arg4);
+ *		baz();
+ *		return res;
+ *	}
+ *
+ * @note __attribute__((naked)) is not mandatory, but without it, GCC gives a
+ *       warning for functions with a return value. It also saves a bit of space
+ *       since it removes a little code that is not necessary.
+ *
+ * See @ref WRAP_FUNC_RAW for more information.
+ */
+#define WRAP_FUNC_STACK_ARGS(preface, name, postface) \
+	WRAP_FUNC_RAW(preface, name, postface, "str lr, %0" :: "m" (lr_bak), \
```

Comment:
I'd assume "bak" stands for backup, or?
Confused a bit, so please choose a more descriptive name